### ISSUE TITLE
Error class for reading mailbox or folders of client

### ISSUE THREAD

**Issue by Firegodbr** [O365/python-o365#892]
I'm using Django to try to connect to the Microsoft API but is not working properly when I called it. The issue being adding this None to the Url when building it. I wish to what cause it. How to go around or is my mistake, It works well when only using:

`account.mailbox().inbox_folder()`

Error:
400 Client Error: Bad Request for url: https://graph.microsoft.com/v1.0/meNone?%24filter=displayName+eq+%27Inbox%27&%24top=1 | Error Message: Resource not found for the segment 'meNone'.

```
token_backend = SqlTokenJsonSave(user=request.user)
        account = Account(credentials=(settings.MICROSOFT_GRAPH_CLIENT_ID, settings.MICROSOFT_GRAPH_CLIENT_SECRET), token_backend=token_backend)
        if account.is_authenticated:
            mailbox = account.mailbox().get_folder(folder_name="Inbox")
            print(mailbox)
```
```
Traceback (most recent call last):
  File "o365\env\lib\site-packages\django\core\handlers\exception.py", line 55, in inner
    response = get_response(request)
  File "o365\env\lib\site-packages\django\core\handlers\base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "o365\mail\views.py", line 68, in read_email
    mailbox = account.mailbox().get_folder(folder_name="Inbox")
  File "o365\env\lib\site-packages\O365\mailbox.py", line 294, in get_folder
    response = self.con.get(url, params=params)
  File "o365\env\lib\site-packages\O365\connection.py", line 847, in get
    return self.oauth_request(url, 'get', params=params, **kwargs)
  File "o365\env\lib\site-packages\O365\connection.py", line 836, in oauth_request
    return self._internal_request(self.session, url, method, **kwargs)
  File "o365\env\lib\site-packages\O365\connection.py", line 798, in _internal_request
    raise HTTPError('{} | Error Message: {}'.format(e.args[0], error_message), response=response) from None
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://graph.microsoft.com/v1.0/meNone?%24filter=displayName+eq+%27Inbox%27&%24top=1 | Error Message: Resource not found for the segment 'meNone'.
```
Edit:
After searching in the code for while:
I found the issue is with this lines: starting after line 261 of mailbox.py

```
    def get_folder(self, *, folder_id=None, folder_name=None):
        """ Get a folder by it's id or name

        :param str folder_id: the folder_id to be retrieved.
         Can be any folder Id (child or not)
        :param str folder_name: the folder name to be retrieved.
         Must be a child of this folder.
        :return: a single folder
        :rtype: mailbox.Folder or None
        """
        if folder_id and folder_name:
            raise RuntimeError('Provide only one of the options')

        if not folder_id and not folder_name:
            raise RuntimeError('Provide one of the options')

        if folder_id:
            # get folder by it's id, independent of the parent of this folder_id
            url = self.build_url(
                self._endpoints.get('get_folder').format(id=folder_id))
            params = None
        else:
            # get folder by name. Only looks up in child folders.
            if self.root:
                url = self.build_url(self._endpoints.get('root_folders'))
```
self._endpoints.get('root_folders') is returning None instead of "/mailFolders". Is there a way to fix it?

I found that's happing when looking into self._endpoints it returns:
`{'settings': '/mailboxSettings'}`

The same one declared here:
```
class MailBox(Folder):
    folder_constructor = Folder
    _endpoints = {
        'settings': '/mailboxSettings',
    }
```
Instead of 
```
class Folder(ApiComponent):
    """ A Mail Folder representation """

    _endpoints = {
        'root_folders': '/mailFolders',
        'child_folders': '/mailFolders/{id}/childFolders',
        'get_folder': '/mailFolders/{id}',
        'root_messages': '/messages',
        'folder_messages': '/mailFolders/{id}/messages',
        'copy_folder': '/mailFolders/{id}/copy',
        'move_folder': '/mailFolders/{id}/move',
        'message': '/messages/{id}',
    }
```
Is possible to fix it?

Edit
Another error found is in the getFolder of Folder(ApiComponent), here:
```
    def get_folder(self, *, folder_id=None, folder_name=None):
        """ Get a folder by it's id or name

        :param str folder_id: the folder_id to be retrieved.
         Can be any folder Id (child or not)
        :param str folder_name: the folder name to be retrieved.
         Must be a child of this folder.
        :return: a single folder
        :rtype: mailbox.Folder or None
        """
        if folder_id and folder_name:
            raise RuntimeError('Provide only one of the options')

        if not folder_id and not folder_name:
            raise RuntimeError('Provide one of the options')

        if folder_id:
            # get folder by it's id, independent of the parent of this folder_id
            url = self.build_url(
                self._endpoints.get('get_folder').format(id=folder_id))
            params = None
        else:
            # get folder by name. Only looks up in child folders.
            if self.root:
                url = self.build_url(self._endpoints.get('root_folders'))
                print(self._endpoints.get('root_folders'))
            else:
                url = self.build_url(
                    self._endpoints.get('child_folders').format(
                        id=self.folder_id))
            params = {'$filter': "{} eq '{}'".format(self._cc('displayName'),
                                                     folder_name), '$top': 1}
        print(url)
        print(self.folder_id)
        response = self.con.get(url, params=params)
        if not response:
            return None

        if folder_id:
            folder = response.json()
        else:
            folder = response.json().get('value')
            folder = folder[0] if folder else None
            if folder is None:
                return None

        self_class = getattr(self, 'folder_constructor', type(self))
        # Everything received from cloud must be passed as self._cloud_data_key
        # We don't pass parent, as this folder may not be a child of self.
        return self_class(con=self.con, protocol=self.protocol,
                          main_resource=self.main_resource,
                          **{self._cloud_data_key: folder})
```
this
```
url = self.build_url(
                    self._endpoints.get('child_folders').format(
                        id=self.folder_id))
```
It doesn't take into account that folder_id can be None so it happend to be None when searching by folder_name

**Comment by oleg-savko** [comment]
also has this problem

**Comment by janscas** [comment]
This was fixed already in PR #885

**Comment by janscas** [comment]
Just released 2.0.23
